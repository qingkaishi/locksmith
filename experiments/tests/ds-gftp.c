/* MWH: This example is based on a quick run-through of gftp.  The
   actual program is pretty tricky because it uses gtk's event-based
   libraries to get things going.  Thus in the real program, functions
   are called from the GTK library, which we presumably don't have
   access to (or else we'll need to analyze it too).  The real structs
   are also more complicated.
*/

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct gftp_request_tag 
{
  unsigned int cancel : 1,
               stopable : 1;
  void *protocol_data;
} gftp_request;

typedef struct gftp_transfer_tag
{
  gftp_request * fromreq;
  unsigned int started : 1,
               cancel : 1,
               ready : 1,
               done : 1;
  int numfiles;
  pthread_mutex_t statmutex,  structmutex;
} gftp_transfer;

gftp_transfer *
gftp_tdata_new (void)
{
  gftp_transfer * tdata;
  tdata = malloc (sizeof (*tdata));
  pthread_mutex_init (&tdata->statmutex, NULL);
  pthread_mutex_init (&tdata->structmutex, NULL);
  return (tdata);
}

static void
cancel_get_trans_password (gftp_transfer * tdata)
{
  if (tdata->fromreq->stopable == 0)
    return;

  pthread_mutex_lock (&tdata->structmutex);
  if (tdata->started)
    {
      tdata->cancel = 1;
      tdata->fromreq->cancel = 1;
    }
  else
    tdata->done = 1;

  tdata->fromreq->stopable = 0;
  tdata->fromreq->protocol_data = NULL;
  pthread_mutex_unlock (&tdata->structmutex);
}

static void
update_file_status (gftp_transfer * tdata)
{
  int n;

  pthread_mutex_lock (&tdata->statmutex);
  n = tdata->numfiles;

  if (n < 1)
    {
      pthread_mutex_unlock (&tdata->statmutex);
      return;
    }

  tdata->numfiles = 5;
  pthread_mutex_unlock (&tdata->statmutex);
}


static int doit1(gftp_transfer *tdata) {
  cancel_get_trans_password(tdata);
  return 1;
}

static int doit2(gftp_transfer *tdata) {
  update_file_status(tdata);
  return 1;
}


int main() {
  int i = 1;
  pthread_t t1, t2;

  while (1) {
    gftp_transfer *tdata = gftp_tdata_new();
    tdata->numfiles = 1; // thread local at the moment

    pthread_create(&t1, NULL, doit1, tdata);
    pthread_create(&t2, NULL, doit2, tdata);
  }
}


