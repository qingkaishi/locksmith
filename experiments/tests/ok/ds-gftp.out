************************* STARTING *************************

shared: alloc.cancel:ds-gftp.c:35
        alloc.done:ds-gftp.c:35
        alloc.numfiles:ds-gftp.c:35

Warning: Possible data race: alloc.numfiles:ds-gftp.c:35 is not protected!
 references:
  dereference of *tdata.numfiles:ds-gftp.c:62 at ds-gftp.c:67
    alloc.numfiles:ds-gftp.c:35 => tdata->numfiles:ds-gftp.c:31
                                => *tdata.numfiles:ds-gftp.c:96 at gftp_tdata_new ds-gftp.c:96
                                => tdata->numfiles:ds-gftp.c:85 at doit2 ds-gftp.c:100
                                => *tdata.numfiles:ds-gftp.c:62 at update_file_status ds-gftp.c:86
  locks acquired:
    concrete tdata->statmutex::-1 -> gftp_tdata_new ds-gftp.c:96
    tdata->statmutex:ds-gftp.c:96
  in: main at ds-gftp.c:91  -> ds-gftp.c:86

  dereference of *tdata.numfiles:ds-gftp.c:62 at ds-gftp.c:75
    alloc.numfiles:ds-gftp.c:35 => tdata->numfiles:ds-gftp.c:31
                                => *tdata.numfiles:ds-gftp.c:96 at gftp_tdata_new ds-gftp.c:96
                                => tdata->numfiles:ds-gftp.c:85 at doit2 ds-gftp.c:100
                                => *tdata.numfiles:ds-gftp.c:62 at update_file_status ds-gftp.c:86
  locks acquired:
    concrete tdata->statmutex::-1 -> gftp_tdata_new ds-gftp.c:96
    tdata->statmutex:ds-gftp.c:96
  in: main at ds-gftp.c:91  -> ds-gftp.c:86

  dereference of *tdata.numfiles:ds-gftp.c:96 at ds-gftp.c:97
    alloc.numfiles:ds-gftp.c:35 => tdata->numfiles:ds-gftp.c:31
                                => *tdata.numfiles:ds-gftp.c:96 at gftp_tdata_new ds-gftp.c:96
  locks acquired:
    <empty>
  in: main at ds-gftp.c:91 


alloc.cancel:ds-gftp.c:35 is protected by:
  concrete tdata->structmutex::-1 -> gftp_tdata_new ds-gftp.c:96
  tdata->structmutex:ds-gftp.c:96
alloc.done:ds-gftp.c:35 is protected by:
  concrete tdata->structmutex::-1 -> gftp_tdata_new ds-gftp.c:96
  tdata->structmutex:ds-gftp.c:96
*************************** DONE ***************************
